### PR TITLE
Add gelu composite rule

### PR DESCRIPTION
**Pull request by GGBond8488**
### PR types
Others 

### PR changes
 Others

### Describe
add composite rule for gelu

### CODE REVIEW

File: python/paddle/incubate/autograd/composite_rules.py [hunk 1]
```diff
@@ -97,3 +97,55 @@ def composite_batchnorm(
...
+
+@REGISTER_COMPOSITE('gelu')
+def gelu_composite(x, approximate):
+    """define composite rule of op gelu"""
+    GELU_CONSTANT = 0.044715
+    M_2_PI = 0.63661977236758134308     #/* 2/pi */. copy from gelu-kernel.cc
+    M_SQRT1_2 = 0.70710678118654752440	#/* 1/sqrt(2) */ copy from gelu-kernel.cc
+
+    # gelu(x) = 0.5 * x * (1 + tanh(sqrt(2 / \pi) * (x + 0.044715 * x^{3})))
+    if approximate:
```
Comment (cyber-pioneer) [inline python/paddle/incubate/autograd/composite_rules.py:111]: 加减乘除可以简化成符号，提升可读性，可以参考bn的case

Comment (GGBond8488) [reply python/paddle/incubate/autograd/composite_rules.py:109]: 简化了加减乘除运算

Comment (GGBond8488) [reply python/paddle/incubate/autograd/composite_rules.py:111]: 简化了加减乘除运算

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                    attrs.set_shape(t)
+                    self.compare_backward()
+
+
+class TestCompositeGeluPrimBackward(unittest.TestCase):
+    "test composite gelu and prim backward"
+
+    def setUp(self):
+        core._set_prim_backward_enabled(True)
+        self.dtypes = ["float32", "float64"]
```
Comment (cyber-pioneer) [inline python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py]: 补充一下float16

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+    gradients = paddle.grad(res, inputs)
+    return gradients
+
+
+class TestCompositeGelu(unittest.TestCase):
+    def setUp(self):
+        core._set_prim_backward_enabled(True)
+        self.dtypes = ["float32", "float64"]
+        self.shapes = [[2, 3, 4], [2, 3]]
```
Comment (cyber-pioneer) [inline python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py]: 建议把一个shape调大些，比如[16, 16, 64,64]

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+
+class TestCompositeGelu(unittest.TestCase):
+    def setUp(self):
+        core._set_prim_backward_enabled(True)
+        self.dtypes = ["float16", "float32", "float64"]
+        self.shapes = [[16, 16, 64, 64], [2, 3, 4], [2, 3]]
+        self.approximates = [True, False]
+
+    def cal_composite_grad(self, inputs):
```
Comment (cyber-pioneer) [inline python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py:83]: only prim forward

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            for j in self.dtypes:
+                for t in self.shapes:
+                    attrs.set_approximate(i)
+                    attrs.set_dtype(j)
+                    attrs.set_shape(t)
+                    self.compare_backward()
+
+
+class TestCompositeGeluPrimBackward(unittest.TestCase):
+    "test composite gelu and prim backward"
```
Comment (cyber-pioneer) [inline python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py]: test composite gelu: prim forward and backward

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+    def setUp(self):
+        core._set_prim_backward_enabled(True)
+        self.dtypes = ["float32", "float64"]
+        self.shapes = [[2, 3, 4], [2, 3]]
+        self.approximates = [True, False]
+
+    def cal_composite_grad(self, inputs):
+        paddle.enable_static()
```
Comment (cyber-pioneer) [inline python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py:156]: core._set_prim_all_enabled(True)

**Comment by cyber-pioneer** [comment]
LGTM

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py (hunk 6 above)
Comment (PaddlePaddle-Gardener) [reply python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py:156]: 增加float16 以及高维的test case

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+
+class PrimeNet(paddle.nn.Layer):
+    def __init__(self, approximate):
+        super(PrimeNet, self).__init__()
+        self.fc = paddle.nn.Linear(4, 4)
+        self.approximate = approximate
+
+    def forward(self, x):
+        # y = self.fc(x)
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py:51]: why not add linear in?

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py (hunk 6 above)
Comment (GGBond8488) [reply python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py:156]: 已修改

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py (hunk 5 above)
Comment (GGBond8488) [reply python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py]: 已修改

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py (hunk 3 above)
Comment (GGBond8488) [reply python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_gelu_grad.py]: 增加 float16 以及 高维case

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        paddle.seed(2022)
+        net = PrimeNet(approximate)
+        sgd = paddle.optimizer.SGD(
+            learning_rate=0.1, parameters=net.parameters()
+        )
+        core._set_prim_all_enabled(use_prim)
+        if use_prim:
+            net = apply_to_static(net, use_prim)
+
+        out = net(data)
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py]: train more steps

**Review by JiabinYang — COMMENTED** [review]
some comments

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py (hunk 7 above)
Comment (GGBond8488) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py:51]: linear dont support float16

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        self.check_prim(net, use_prim)
+
+        return out.numpy()
+
+    def check_prim(self, net, use_prim):
+        if not use_prim:
+            return
+        fwd_ops = [op.type for op in net.forward.main_program.block(0).ops]
+        # Ensure that gelu is splitted into small ops
+        self.assertTrue('gelu' not in fwd_ops)
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py:98]: is there any _grad used in gelu grad？

File: python/paddle/incubate/autograd/composite_rules.py [hunk 10]
```diff
@@ -97,3 +97,27 @@ def composite_batchnorm(
...
+        # gelu(x) = 0.5 * x * (1 + tanh(sqrt(2 / \pi) * (x + 0.044715 * x^{3})))
+        kAlpha = full(x.shape, M_2_SQRTPI * M_SQRT1_2, x.dtype)
+        GELU_CONSTANT = full(x.shape, 0.044715, x.dtype)
+        tanh_out = tanh(kAlpha * (x + GELU_CONSTANT * pow(x, 3)))
+        out = x * half * (one + tanh_out)
+        return out
+
+    else:
+        # gelu(x) = 0.5 * x *  (1 + erf(x / sqrt(2)))
+        cdf = half * (one + erf(x * full(x.shape, M_SQRT1_2, x.dtype)))
```
Comment (JiabinYang) [inline python/paddle/incubate/autograd/composite_rules.py:121]: is erf a primitive?

**Review by JiabinYang — COMMENTED** [review]
more comments

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py (hunk 9 above)
Comment (GGBond8488) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py:98]: 'pow_grad' and 'erf_grad'

File: python/paddle/incubate/autograd/composite_rules.py (hunk 10 above)
Comment (GGBond8488) [reply python/paddle/incubate/autograd/composite_rules.py:121]: yes

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py (hunk 8 above)
Comment (GGBond8488) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py]: will add 10 steps

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py (hunk 9 above)
Comment (GGBond8488) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_cinn_prim_gelu.py:98]: pow_grad has been circumvented, erf_grad see https://github.com/PaddlePaddle/Paddle/pull/50436

**Review by JiabinYang — APPROVED** [review]
LGTM

**Pull request merged**